ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+
+    # Block execution until both promises are resolved
+    activity_result, workflow_result = async.all(activity_promise, workflow_promise).await
+
+    activity_result # => 3
+    workflow_result # => 'Hello, parent!'
+    
+    promise = async { workflow.execute_activity(SumActivity, 3, 4) }
+    
+    # individual promises can also be waited on
+    result = promise.await
+    
+    result # => 7
+  end
+end
+```
+
+A bit more involved example:
+
+```ruby
+class AnotherAsyncWorkflow < Temporal::Workflow
+  def execute
+    promise_1 = async { wait_and_sum(1, 2, delay: 15) }
+    promise_2 = async { wait_and_sum(3, 4, delay: 10) }
+
+    # The same exact method can be called in a blocking fashion
+    wait_and_sum(5, 6, delay: 5) # => 11
+
+    # Block execution until any promise is resolved
+    async.any(promise_1, promise_2).await
+
+    # Returns immediately because it will already be resolved (shorter delay)
+    promise_2.result # => 7
+
+    # This would block until the promise is resolved (longer delay)
+    promise_1.result # => 3
+  end
+
+  private
+
+  def wait_and_sum(a, b, delay:)
+    workflow.sleep(delay)
+    workflow.execute_activity(SumActivity, a, b)
+  end
+end
+```
+
+Methods `async.all` and `async.any` return promises themselves, so they can be used just like the
+promises returned by an `async {}` call.
+
+Also worth noting that `async {}` calls can be nested to model more complex business logic.
+
+
+## Activity heartbeats & cancellations
+
+This approach uses `Thread#raise` that will eject the execution out of the current context. In cases
+where this is not acceptable we will allow to either wrap a critical code in an `activity.shield`
+block (that will not raise until the block has executed) or opt-out of this default behaviour
+altogether and use `activity.cancelled?` flag to determine when an activity was cancelled.
+
+```ruby
+class LongRunningActivity < Temporal::Activity
+  def execute
+    10.times do
+      sleep 1
+      activity.heartbeat('ping')
+    end
+  rescue Temporal::Error::ActivityCancelled
+    raise 'this activity is cancelled'
+  end
+end
+```
+
+Here is what it will look like from a workflow perspective:
+
+```ruby
+class InnerCancelWorkflow < Temporal::Workflow
+  def execute
+    # the block supplies an optional argument
+    async do |outer_promise|
+      async do
+        workflow.sleep(30)
+        outer_promise.cancel
+      end
+
+      # One of these blocking activities will end up getting cancelled 
+      #   unless they all finish within 30 seconds
+      workflow.execute_activity(LongRunningActivity)
+      workflow.execute_activity(LongRunningActivity)
+      workflow.execute_activity(LongRunningActivity)
+    end.await
+  end
+end
+```
+
+#### Alternative approaches
+
+- Overall using `Thread#raise` is considered bad practice because it might raise in the
+  middle of some process. We might want to consider opting in to this behaviour instead
+
+
+## Workflow cancellations
```

Comment:
Not sure where to bring this up, but kinda related, the default for temporal-ruby when the workflow implementation throws is to fail the workflow, whereas in Java, the workflow keeps retrying unless you specifically throw a fatal error.  
I'm guessing we'll switch since we're moving to rust-core but wanted to double check.  I would be in favor of defaulting to retrying since that unlocks the ability to deploy code to fix workflows and seems more inline with temporal's ethos.